gtk: add import/export memory menu items

prg318 opened the pull request:
This pull request is a rebase of https://github.com/TASVideos/desmume/pull/347 .  When I originally submitted this, there was a lot of recent changes to the gtk code in desmume.  I finally got around to rebasing this against the latest gtk changes in desmume

This branch adds "Import Backup Memory" and "Export Backup Memory" entries to the GTK File menu.  Previously, this functionality was only available on the win32 version.

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,131 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+    gtk_file_filter_set_name(pFilter_any, "All supported types");
+
+    /* Creating the selection window */
+    pFileSelection = gtk_file_chooser_dialog_new("Import Backup Memory From ...",
```
Please use [`GtkFileChooserNative`](https://developer.gnome.org/gtk3/unstable/gtk3-GtkFileChooserNative.html) instead, so that it can use the user’s preferred file picker instead of forcing the GTK one, plus it will work from inside sandboxes like Flatpak’s.

You can have a look around, for instance [`PlayMovieDialog()`](https://github.com/TASVideos/desmume/blob/master/desmume/src/frontend/posix/gtk/main.cpp#L1743) does pretty much the same thing.

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,131 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+    pFileSelection = gtk_file_chooser_dialog_new("Import Backup Memory From ...",
+            GTK_WINDOW(pParent),
+            GTK_FILE_CHOOSER_ACTION_OPEN,
+            GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL,
```
Please also avoid using `GtkStock` symbols, as those are deprecated.

I tried to avoid having any deprecated symbol to ease GTK 4 porting.  I already have a [working branch](https://github.com/linkmauve/desmume/tree/gtk4) for it, I’m just waiting for it to get a proper release before switching.

Maybe it’d be better to `#define GTK_DISABLE_DEPRECATED` somewhere, to avoid people adding new deprecated symbols.

linkmauve replied:
There, as of 64da66f37b245dccac694096b47648af16dd2339 using deprecated symbols will be an error instead of a warning.

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,131 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+    gtk_file_chooser_add_filter(GTK_FILE_CHOOSER(pFileSelection), pFilter_any);
+
+    /* Showing the window */
+    switch(gtk_dialog_run(GTK_DIALOG(pFileSelection))) {
```
I’ve also switched those `switch` statements with `if` ones, since they have a single branch anyway.

Note that if you upgrade to `GtkFileChooserNative` you’ll have to check `GTK_RESPONSE_ACCEPT` instead of `GTK_RESPONSE_OK`.

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,131 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+    gtk_file_filter_set_name(pFilter_any, "All supported types");
+
+    /* Creating the selection window */
+    pFileSelection = gtk_file_chooser_dialog_new("Export Backup Memory To ...",
```
Same here.

linkmauve reviewed (changes requested):
Aside from the requested changes, LGTM!

I haven’t tested this PR, only reviewed it.

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,115 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+    pFileSelection = gtk_file_chooser_native_new("Import Backup Memory From ...",
+            GTK_WINDOW(pWindow),
+            GTK_FILE_CHOOSER_ACTION_OPEN,
+	    "_Open", "_Cancel");
```
You have a tabulation here, please replace it with spaces.

prg318 replied:
fixed in eec3dd2d9cdda2bbaed76873045b0c822225b657

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,115 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+    pFileSelection = gtk_file_chooser_native_new("Export Backup Memory To ...",
+            GTK_WINDOW(pWindow),
+            GTK_FILE_CHOOSER_ACTION_SAVE,
+	    "_Save", "_Cancel");
```
Here too.

prg318 replied:
fixed in eec3dd2d9cdda2bbaed76873045b0c822225b657

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,115 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+    /* Showing the window */
+    int response = gtk_native_dialog_run(GTK_NATIVE_DIALOG(pFileSelection));
+    if (response == GTK_RESPONSE_ACCEPT) {
+	GFile *file = gtk_file_chooser_get_file(GTK_FILE_CHOOSER(pFileSelection));
```
And here too.

prg318 replied:
fixed in eec3dd2d9cdda2bbaed76873045b0c822225b657

linkmauve reviewed (changes requested):
Aside from these three nits, it looks cool!

How can I test this feature?

prg318 commented:
Thanks for the feedback!  I went ahead and pushed a commit to:
- Use gtk_native_dialog
- Removed gtk stock symbols
- Changed the single case switches into if statements

I also added code to enable and disable the menu items based on whether the emulator is running (similar to print screen and other menu items).  I also modified the casing of the menu items to be consistent with the other items in the file menu.

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,115 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+            "_Open", "_Cancel");
+
+    /* Only the dialog window is accepting events: */
+    gtk_window_set_modal(GTK_WINDOW(pFileSelection), TRUE);
```
You can’t cast a `GtkFileChooserNative` into a `GtkWindow`, it isn’t even necessarily a `GtkWidget` as it could be implemented by a completely foreign toolkit, possibly even out-of-process.

This prints a critical warning when running it.

prg318 replied:
Thanks; I've removed this in e036298aa2dc0903bdd2417f49f1de7e35cfea5f

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,115 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+    gtk_file_chooser_set_do_overwrite_confirmation (GTK_FILE_CHOOSER (pFileSelection), TRUE);
+
+    /* Only the dialog window is accepting events: */
+    gtk_window_set_modal(GTK_WINDOW(pFileSelection), TRUE);
```
Same here.

prg318 replied:
also removed in  e036298aa2dc0903bdd2417f49f1de7e35cfea5f

prg318 commented:
> Aside from these three nits, it looks cool!
> 
> How can I test this feature?
Thanks!
Here's how I've been testing:
1) Download this pokemon white save file:  https://xannode.com/whitetest100.sav
2) Open desmume
3) Open a backup dump of pokemon white
4) Select "file -> import backup memory" and select whitetest100.sav
5) The emulation should restart.  On the load file screen, you should have a file with 0 badges but a complete pokedex.

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,115 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+        gchar *sPath = g_file_get_path(file);
+
+        if(MMU_new.backupDevice.exportData(sPath) == false ) {
+            GtkWidget *pDialog = gtk_message_dialog_new(GTK_WINDOW(pFileSelection),
```
Same here.

prg318 replied:
addressed in 8230656fd6cacabdbbe240cd67f6f728b01be4cf

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,115 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+        gchar *sPath = g_file_get_path(file);
+
+        if(MMU_new.backupDevice.importData(sPath) == false ) {
+            GtkWidget *pDialog = gtk_message_dialog_new(GTK_WINDOW(pFileSelection),
```
Same here.

prg318 replied:
addressed in 8230656fd6cacabdbbe240cd67f6f728b01be4cf

prg318 commented:
If I recall correctly, I originally based ExportSaveDialog and ImportSaveDialog on an older version of the SaveStateDialog().  I should have done a more thorough job of rebasing after SaveStateDialog() was improved/updated a while back

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,109 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+    int response = gtk_native_dialog_run(GTK_NATIVE_DIALOG(pFileSelection));
+    if (response == GTK_RESPONSE_ACCEPT) {
+        GFile *file = gtk_file_chooser_get_file(GTK_FILE_CHOOSER(pFileSelection));
+        gchar *sPath = g_file_get_path(file);
```
In order to avoid getting a “Unable to export” error, add a .sav extension if the user didn’t type one. I got this error when trying to save it as `/tmp/foo` without extension.

prg318 replied:
Good call.  This should be addressed in ecd600e76f06e631c15748d41b46b4702f253b50

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,109 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+                    GTK_DIALOG_MODAL,
+                    GTK_MESSAGE_ERROR,
+                    GTK_BUTTONS_OK,
+                    "Unable to export :\n%s", sPath);
```
Extra space before “:” isn’t proper English.

prg318 replied:
aba5d0fd3c11b9cc6ada5c1573dc0ab76cf3b40a

linkmauve commented on desmume/src/frontend/posix/gtk/main.cpp:
```diff
@@ -1779,6 +1795,109 @@ static void PlayMovieDialog(GSimpleAction *action, GVariant *parameter, gpointer
+                    GTK_DIALOG_MODAL,
+                    GTK_MESSAGE_ERROR,
+                    GTK_BUTTONS_OK,
+                    "Unable to import :\n%s", sPath);
```
Same here.

linkmauve reviewed (approved):
I don’t have my Pokémon White cartridge dumped atm, but from a cursory test on another game it seems to work.

Good work, thanks!

prg318 commented:
@linkmauve Thank you for the thorough pull request and thanks for providing the quick tips to help me address your comments.  I really appreciate your time and caring!

linkmauve commented:
You’re welcome. :)
